# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from tqdm import tqdm

In [2]:
boston = load_boston()
x = boston.data #independent variables
y = boston.target #target variable
print(x.shape, y.shape)

(506, 13) (506,)


In [3]:
import pandas as pd
pd.DataFrame(x, columns=boston.feature_names).head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [4]:
def generating_samples(input_data, target_data):
    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    selecting_rows = np.random.choice(506, 303, replace=False)
    replacing_rows = np.random.choice(303, 203)
    selecting_columns = np.random.choice(11, np.random.choice(9, 1)+3, replace=False)+2
    
    sample_data = input_data[selecting_rows[:, None] , selecting_columns]
    target_of_sample_data = target_data[selecting_rows]
    
#     Replicating Data
    replicated_sample_data = sample_data[replacing_rows]
    target_of_replicated_sample_data = target_data[replacing_rows]
    
#     Concatenating Data
    final_smaple_data = np.vstack((sample_data, replicated_sample_data))
    final_target_data = np.hstack((target_of_sample_data, target_of_replicated_sample_data))
    
    return list(final_smaple_data), list(final_target_data), list(selecting_rows), list(selecting_columns)

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists


<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [5]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [6]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data = []
list_output_data = []
list_selected_row = []
list_selected_columns = []

for i in range(30):
    a,b,c,d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [7]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [8]:
from sklearn.tree import DecisionTreeRegressor

def Models(list_input_data, list_output_data):
    all_models = []
    for i in range(30):
        model = DecisionTreeRegressor().fit(list_input_data[i], list_output_data[i])
        all_models.append(model)
    return all_models

list_of_all_models = Models(list_input_data, list_output_data)
# print(list_of_all_models)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [9]:
def MSE(list_of_all_models, list_input_data, list_output_data):
    array_of_Y = [ list_of_all_models[i].predict(list_input_data[i]) for i in range(30) ]
    mse = np.mean([mean_squared_error(array_of_Y[i], list_output_data[i]) for i in range(30)])
    return mse

mse = MSE(list_of_all_models, list_input_data, list_output_data)
print('mse = ', mse)

mse =  29.166878592195392


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [14]:
def OOB_Score(x, y, list_of_all_models, list_selected_row, list_selected_columns):
    
    list_oob_rows = [list( set(range(506))-set(list_selected_row[i]) ) for i in range(30)]
    
    list_oob_input_data = []
    list_oob_output_data = []
    for i in range(30):
        oob_data = x[list_oob_rows[i], : ]
        oob_data = oob_data[ : , list_selected_columns[i]]
        list_oob_input_data.append(oob_data)
        list_oob_output_data.append(y[list_oob_rows[i]])
        
    list_oob_output_data_predicted = []
    for i in range(30):
        error = []
        for j in range(203):
            err = []
            for k in range(30):
                if list_oob_rows[i][j] not in list_selected_row[k]:
                    oob_data = x[list_oob_rows[i][j], list_selected_columns[k]]
                    err.append(list_of_all_models[k].predict(oob_data.reshape(1,-1))[0])
            error.append(np.mean(err))
        list_oob_output_data_predicted.append(error)
    oob_score = np.median([mean_squared_error(list_oob_output_data_predicted[i], list_oob_output_data[i]) for i in range(30)])
    return(oob_score)
    
oob_score = OOB_Score(x, y, list_of_all_models, list_selected_row, list_selected_columns)
print('oob score = ', oob_score)

oob score =  22.81217126632297


# <font color='blue'><b>Task 2</b></font>
##### Computing CI of OOB Score and Train MSE

In [15]:
def ConfidenceInterval(score):
    mean = np.mean(score)
    std = np.std(score)
    ci = ( mean-2*std/np.sqrt(len(score)), mean+2*std/np.sqrt(len(score)) )
    return ci

In [16]:
mse_35 = []
oob_score_35 = []

for i in tqdm(range(35)):
    list_input_data = []
    list_output_data = []
    list_selected_row = []
    list_selected_columns = []
    for i in range(30):
        a,b,c,d = generating_samples(x, y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
    
    list_of_all_models = Models(list_input_data, list_output_data)
    
    mse = MSE(list_of_all_models, list_input_data, list_output_data)
    mse_35.append(mse)
    oob_score = OOB_Score(x, y, list_of_all_models, list_selected_row, list_selected_columns)
    oob_score_35.append(oob_score)

print('95% CI of mse: ', ConfidenceInterval(mse_35))
print('95% CI of obb score: ', ConfidenceInterval(oob_score_35))

100%|██████████████████████████████████████████████████████████████████████████████████| 35/35 [09:05<00:00, 15.30s/it]


95% CI of mse:  (28.363018905163802, 28.939696928118455)
95% CI of obb score:  (23.119001089549734, 24.40445901616865)


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [17]:
xq = np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])

y_xq_pred = np.median([ list_of_all_models[i].predict(xq[list_selected_columns[i]].reshape(1,-1)) for i in range(30)])
print(y_xq_pred)

19.424999999999997


<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

#### Task 1:
mean square error is found to be 29.17 <br>
and out of bag score is found to be 22.81

#### Task 2:
95% CI of mse is:  (28.36, 28.94) <br>
95% CI of obb score is:  (23.12, 24.40)

#### Task 3:
predicted output of given query point xq=[0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] <br>
is 19.42
